app/services/process_mentions_service.rb

```diff
@@ -10,23 +10,21 @@ class ProcessMentionsService < BaseService
   def call(status)
     return unless status.local?
 
-    status.text.scan(Account::MENTION_RE).each do |match|
-      username, domain  = match.first.split('@')
-      mentioned_account = Account.find_remote(username, domain)
-
-      if mentioned_account.nil? && !domain.nil?
-        begin
-          mentioned_account = resolve_remote_account_service.call(match.first.to_s)
-        rescue Goldfinger::Error, HTTP::Error
-          mentioned_account = nil
-        end
+    status.text = status.text.gsub(Account::MENTION_RE) do |match|
+      begin
+        mentioned_account = resolve_remote_account_service.call($1)
+      rescue Goldfinger::Error, HTTP::Error
+        mentioned_account = nil
       end
 
-      next if mentioned_account.nil?
+      next match if mentioned_account.nil?
 
       mentioned_account.mentions.where(status: status).first_or_create(status: status)
+      mentioned_account.domain.nil? ? "@#{mentioned_account.username}" : "@#{mentioned_account.username}@#{mentioned_account.domain}"
     end
 
+    status.save!
```

Comment:
at this point have we already delivered the message to subscribers? I don't know if we're allowed to re-write the message here then—we'd have to pull this code somewhere else.

Reply:
No we haven't, this is right before delivering it.

Reply:
How do we know that though? I don't know if this is the right thing to do from a code organization perspective. adding a transaction might create a nested transactions error, for example.

Reply:
I understand your concern, and I share it to some extent. However, the serialized status sent over both the WebUI and remote servers through OStatus and ActivityPub already depends on this function, as the serialization depends on the `mentioned_account.mentions` created by this function.